frb_codegen/src/config/opts.rs

```diff
@@ -29,17 +43,218 @@ pub struct Opts {
+                        let extra_source_rust_content = read_rust_file(&PathBuf::from(&code_path));
+                        //↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓extra parse↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓↓
+                        let extra_file_ast = syn::parse_file(&extra_source_rust_content).unwrap();
+                        log::debug!("Phase: Parse EXTRA AST to IR");
+                        let extra_ir_file = parser::parse(
+                            &extra_source_rust_content,
+                            extra_file_ast,
+                            &self.manifest_path,
+                            self.block_index,
+                            self.shared,
+                            &[],
+                        )
+                        .unwrap();
+                        log::debug!("Finished Phase: Parse EXTRA AST to IR");
+                        //↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑extra parse↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑
+
+                        // collect extra_ir_file.funcs into vec
+                        let extra_path_funcs =
+                            extra_ir_file.funcs(false).into_iter().collect::<Vec<_>>();
+                        e.insert(extra_path_funcs.clone());
+
+                        extra_path_funcs
+                    } else {
+                        continue;
+                    }
+                } else {
+                    extra_func_map[&raw_code_path].clone()
+                };
+
+                // add types(struct/enum) methods defined in extra rust modules
+                let mut new_funcs = Vec::new();
+                for each_func in extra_path_funcs {
+                    if each_func.name.ends_with(&(format!("__{type_name}"))) {
+                        let mut method = each_func.clone();
+                        if ir_file
+                            .get_shared_type_names(
+                                true,
+                                Option::<Box<dyn Fn(&IrType) -> bool>>::None,
+                            )
+                            .contains(&type_name)
+                        {
+                            method.shared = true;
+                        }
+
+                        if !ir_file.funcs(false).contains(&method) {
+                            new_funcs.push(method.clone());
+                        }
+                    }
+                }
+
+                ir_file.append_funcs(new_funcs);
+
+                // update the ir_file back to the global irfile map
+                IR_FILE_MAP.with(|data| {
+                    let mut ir_file_map = data.borrow_mut();
+                    ir_file_map.insert(self.block_index, ir_file.clone());
+                });
+
+                // if new methods are added, then it is essential to refine (no-)shared types.
+                if original_func_len != ir_file.funcs(false).len() {
+                    ir_file.fetch_shared_types_forcely(all_configs);
+                }
+            }
+        });
```

Comment:
it is a bit hard to understand what's going on here (or maybe it is because after several hours or reviewing I am a bit dizzy). maybe we can 

1. extract the part to a new function (or multi functions, or a file, or whatever)
2. improve the logic a little bit